### PR TITLE
Fix #13 memorised samples are not computed correctly

### LINKED ISSUE #13: Memorization is not as in the paper, ours is very low
We need to figure out why noisy labels aren't being memorized. Right now the basic model and ELR seem very equal, while in the paper, basic ResNet with CE memorized a lot of noisy samples.

image

Our basic CE model shows no signs of memorization (always <1%). I personally dont know what is wrong, we should look at the authors' code.

### PR DESCRIPTION
**Pull request by YinengWang**
Fixes #13 
Compute memorization only on the wrong labels of the training data

### CODE REVIEW

**Comment by alessioGalatolo** [comment]
About the memorization, it is my understanding that it doesn't apply to datasets that don't have explicit noise (CDON's one). Can we make it modular to include this case?

**Comment by YinengWang** [comment]
> About the memorization, it is my understanding that it doesn't apply to datasets that don't have explicit noise (CDON's one). Can we make it modular to include this case?

Yes we should. We make it an option that can be enabled/disabled in the config.

**Review by Alfred-N — COMMENTED** [review]
Great

**Comment by rlindsberg** [comment]
Skærmbillede 2021-05-15 kl  20 36 56

Are you aware of this warning?

**Comment by YinengWang** [comment]
> Skærmbillede 2021-05-15 kl 20 36 56
> 
> Are you aware of this warning?

It's unexpected and I didn't encounter it. It seems to be caused by calling a `wandb.log` with a smaller `step` after a larger one, but it's unlikely to happen, according to the code.

**Comment by Alfred-N** [comment]
I think the bug might be happening because of the following:

Line 119:
`wandb.log({"epoch": epoch, "loss": loss_batch}, step=batch_ct)`

In combination with

Line 134-136.
`            wandb.log({"correct_clean": correct_in_clean, "incorrect_clean": incorrect_in_clean,
                       "correct_wrong": correct_in_wrong, "memorized_wrong": memorized_in_wrong,
                       "incorrect_wrong": incorrect_in_wrong})`

From the wandb documentation:

> Each time you log, we increment the step by default, letting you view metrics over time.

So maybe if we put this in a if-else condition 

```
 if (batch idx == end batch idx)
wandb.log({"correct_clean": correct_in_clean, "incorrect_clean": incorrect_in_clean,
                       "correct_wrong": correct_in_wrong, "memorized_wrong": memorized_in_wrong,
                       "incorrect_wrong": incorrect_in_wrong,"epoch": epoch, "loss": loss_batch}, step=batch_ct)
else:
wandb.log({"epoch": epoch, "loss": loss_batch}, step=batch_ct)

```
it might clear this error.

We could also try including step=epoch Line 134-136. (not sure if multiple steps are allowed)

**Comment by Alfred-N** [comment]
Oh wait, I think

`example_ct += len(inputs)
batch_ct += 1`

has been removed?

**Comment by Alfred-N** [comment]
> Oh wait, I think
> 
> `example_ct += len(inputs) batch_ct += 1`
> 
> has been removed?

Nevermind, it's still there

**Comment by Alfred-N** [comment]
> I think the bug might be happening because of the following:
> 
> Line 119:
> `wandb.log({"epoch": epoch, "loss": loss_batch}, step=batch_ct)`
> 
> In combination with
> 
> Line 134-136.
> ` wandb.log({"correct_clean": correct_in_clean, "incorrect_clean": incorrect_in_clean, "correct_wrong": correct_in_wrong, "memorized_wrong": memorized_in_wrong, "incorrect_wrong": incorrect_in_wrong})`
> 
> From the wandb documentation:
> 
> > Each time you log, we increment the step by default, letting you view metrics over time.
> 
> So maybe if we put this in a if-else condition
> 
> ```
>  if (batch idx == end batch idx)
> wandb.log({"correct_clean": correct_in_clean, "incorrect_clean": incorrect_in_clean,
>                        "correct_wrong": correct_in_wrong, "memorized_wrong": memorized_in_wrong,
>                        "incorrect_wrong": incorrect_in_wrong,"epoch": epoch, "loss": loss_batch}, step=batch_ct)
> else:
> wandb.log({"epoch": epoch, "loss": loss_batch}, step=batch_ct)
> ```
> 
> it might clear this error.
> 
> We could also try including step=epoch Line 134-136. (not sure if multiple steps are allowed)

Or maybe just add step=batch_ct everywhere

**Comment by Alfred-N** [comment]
Fixed

File: main.py [hunk 1]
```diff
@@ -293,7 +293,7 @@ def main():
         classes=10,
         noise_rate=0.2,
         is_symmetric_noise=True,
-        fraction=1.0,
+        fraction=0.01,
```
Comment (rlindsberg) [inline main.py:296]: You need to change this back...

**Review by rlindsberg — CHANGES_REQUESTED** [review]

File: main.py (hunk 1 above)
Comment (Alfred-N) [reply main.py]: It's changed back.

**Review by rlindsberg — APPROVED** [review]
